components/data-table/index.jsx

```diff
@@ -259,6 +259,10 @@ class DataTable extends React.Component {
 	componentDidUpdate() {
 		if (this.props.fixedHeader) {
 			this.resizeFixedHeaders();
+			// If the first page of results isn't enough to allow the user to scroll it causes
+			// the user to get into a state where they cannot load the second page.
+			// Simulating a scroll here will ensure that enough rows are loaded to enable scrolling
+			this.onScrollerScroll();
```

Comment:
Could this have an unwanted effect of triggering an `onLoadMore()` call when the scroll position is outside the `loadMoreOffset` area?  For example, when there is an unrelated prop change like `selection` changes?

Reply:
Is it possible for the consumer to calculate the right number of items and to provide them on initial render?  Or do we need to have DataTable repeatedly request items until there are enough?

Reply:
1) It will call `onScrollerScroll()` when an unrelated prop changes, but it will only trigger `onLoadMore()` if the the table would need to load more given the current scroll position.
2) I think it would be extremely difficult for the consumer to compute the correct number of items to load on the initial render because it would need to reach into the inside of the DataTable to get the height, and then compute the height of each row. That feels like it would be exposing too much inside logic to the outside consumer, creating a more fragile DataTable.

Reply:
You could calculate the height yourself outside of the DataTable using another container surrounding the DataTable to prefetch the appropriate number of items to fill the table.  That way you wouldn't need to know the inside logic.

Reply:
@Prophetess wouldn't you then need to duplicate the entire logic inside of onScrollScrolled in each consumer?

I don't see a case where we wouldn't want this to be the default behavior. Right now, infinite scrolling is broken on larger screens, and I don't see the need to make _every_ consumer to implement this internal logic to prevent this bug.

Reply:
Instead of doing this check in `componentDidUpdate`, what do you think of moving this to `componentDidMount`, and call `onLoadMore` with the number of rows needed to fill the remaining space?  Something like:
```
componentDidMount() {
	if (hasSpaceForMoreRows) {
		const numRows = calculateRowsToFillAvailableSpace();
		this.props.onLoadMore(numRows);
	}
}
```
That way, the consumer knows how many rows are needed to fill the available space; plus, we wouldn't need to do the check on every prop update.

Reply:
I would also prefer it being called only on mount instead of on every update.  Also it would be good to maybe have a prop `loadMoreOnMount` or something for consumers to be able turn on this behavior in case they won't want it called.

Reply:
@garygong I'm not sure moving the logic to `componentDidMount` would actually fix the bug we are having. Some consumers (including us) create the infinite scrolling data table with `items` set to an empty array. We do this because the data we put into the table is fetched asynchronously from an api, but we want to show the user the table with a spinner in the meantime (a perfectly normal UX pattern). Since the array is empty at the beginning, the data table would call the `onLoadMore` prop, however since we are already fetching the first page, another fetch wouldn't be triggered.

here's what the sequence of events would look like
- Start loading data table items from API
- Create infinite scrolling Data Table with `items={[]}`
- Data Table calls `onLoadMore`
- Our app sees we are already loading the first page, do nothing with the `onLoadMore` call since it would be fetching the same data (in fact, this is what the docs recommend <sup>1</sup>)
- Finish loading data table items from API
- Update Data Table with new items
- Nothing??? (unless this current change is implemented)
- User still cannot scroll because there aren't enough items

I also don't believe the `onLoadMore(numRows: number)` would be a good fix for two reasons. First, a lot of API's have a limit on how large of page sizes can be requested. Secondly, we'd still be forcing the consumer of this library to understand why `onLoadMore` has a `numRows` property, and why they need to load at least that many rows. It makes the API for this complex component even more complex.

If y'all are concerned with the the amount of work that is being done in `componentDidUpdate`, then you don't want to look at the `resizeFixedHeaders()` method. It calls `getBoundingClientRect()` in a loop, which in some browsers trigger a layout calculation.

1: `onLoadMore` docs - `This will be called multiple times while the table is being scrolled within the loadMoreOffset. Please track whether or not loading is in progress and check it at the start of this function to avoid executing your callback too many times.`

Reply:
That seems like it could work, but would it really be better than this solution where we can solve it at the library level for everyone automatically? I guess I really don't see the huge hesitation towards adding a small check on `componentDidUpdate` (which won't even be ran if they aren't using fixed headers/infinite scrolling).


> Existing consumers of the API can ignore numRows and would be unaffected

If by unaffected you mean that they'll still have this bug in their application, yeah.

Reply:
Now that I think about the `onLoadMore(numRows: number)`, we couldn't reliably calculate `numRows` for the consumer. We can't compute it because the height of each row is variable. Look at Custom Classes on the documentation page https://react.lightningdesignsystem.com/components/data-tables/

Reply:
Yeah, the hesitation is that DataTable is a heavily used component with a history of performance issues.  So anything we can do to avoid unnecessary work should be done.  One possible solution for handling custom rows with unknown height is to provide the space in pixels, and allow the consumer to decide how to fill that space; and `numRows` would represent the standard row height.

If this seems too involved, we could go with the current check, but to only do it when infinite loading is enabled.  Ideally, there would be a prop for this like `enableInfiniteLoading`, but we can't introduce that now because that would break existing consumers.  Maybe we can use `onLoadMore` being specified as an indicator that infinite loading is enabled?

Reply:
`loadMoreIfNeeded` already does two checks to ensure that unnecessary work isn't done, `if (this.props.hasMore && this.props.onLoadMore) {`. From my understanding of the code and docs, `hasMore` and `onLoadMore` are the props needed to enable infinite scroll.

Reply:
I am also concerned with breaking existing behavior.  When I implemented it I made it so that `onLoadMore` is triggered by a user action: scrolling. So I am not expecting it to call `onLoadMore` on an update... having this called now when it updates will trigger an unnecessary fetch for me.

Reply:
Would it be possible to create a new callback prop instead of reusing the existing `onLoadMore` prop?

Reply:
@Prophetess It won't call `onLoadMore` unless more rows are actually needed to fill the space. So in your case, this change shouldn't impact you. I would appreciate if you could pull this change down to confirm that though.

Reply:
Sure if it only calls `onLoadMore` when needed I'm fine with calling it in the update.  But the documentation for `onLoadMore` should be updated then to let consumers know of this behavior.

Reply:
@Prophetess I've updated the documentation. Please let me know if you can think of better wording.

Reply:
Sorry for the delay, thanks a lot!